src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -154,89 +178,154 @@ export class MongoDBEventStore implements EventStore<number> {
+      });
+    }
+
     // @ts-expect-error The actual `EventType` is different across each stream document,
     // but the collection was instantiated as being `EventStream<Event>`. Unlike `findOne`,
     // `findOneAndUpdate` does not allow a generic to override what the return type is.
     const updatedStream: WithId<EventStream<EventType>> | null =
-      await this.collection.findOneAndUpdate(
-        {
-          streamName: { $eq: streamName },
-          events: { $size: stream.events.length },
-        },
-        {
-          $push: { events: { $each: eventCreateInputs } },
-          $set: { updatedAt: new Date() },
-        },
-        { returnDocument: 'after' },
+      streamPositionDeserializer(
+        await this.collection.findOneAndUpdate(
+          {
+            streamName: { $eq: streamName },
+            'metadata.streamPosition': {
+              $eq: stream.metadata.streamPosition.toString(),
+            },
+          },
+          {
+            $push: { events: { $each: eventCreateInputs } },
+            $set: {
+              'metadata.updatedAt': new Date(),
+              'metadata.streamPosition': (
+                stream.metadata.streamPosition + BigInt(events.length)
+              ).toString(),
+            },
+          },
+          { returnDocument: 'after' },
+        ),
       );
 
     if (!updatedStream) {
-      const currentStream = await this.collection.findOne({
-        streamName: { $eq: streamName },
-      });
+      const currentStream = streamPositionDeserializer(
+        await this.collection.findOne({
+          streamName: { $eq: streamName },
+        }),
+      );
       throw new ExpectedVersionConflictError(
-        currentStream?.events.length ?? -1,
-        stream.events.length,
+        currentStream?.metadata?.streamPosition ?? 0n,
+        stream.metadata.streamPosition,
       );
     }
 
-    const { streamType, streamId } = fromStreamName(streamName);
-
-    await executeProjections(
-      {
-        streamName,
-        streamType,
-        streamId,
-        streamVersion: updatedStream.events.length,
-        events: updatedStream.events,
-      },
-      options?.projections,
-    );
-
     return {
-      nextExpectedStreamVersion: updatedStream.events.length,
+      nextExpectedStreamVersion: updatedStream.metadata.streamPosition,
       createdNewStream,
     };
   }
 }
 
-export const getMongoDBEventStore = (collection: Collection<EventStream>) => {
-  const eventStore = new MongoDBEventStore(collection);
+export const getMongoDBEventStore = (
+  options: ConstructorParameters<typeof MongoDBEventStore>[0],
+) => {
+  const eventStore = new MongoDBEventStore(options);
   return eventStore;
 };
 
-function executeProjections<EventType extends Event>(
-  params: StreamToProject<EventType>,
-  projections?: Array<Projection<EventType>>,
-) {
-  return Promise.all((projections ?? []).map((project) => project(params)));
+async function handleProjections<
```

Comment:
**FIX:** So we have here two types of projections:
1. Projections to the same document (let's call them _"inline"_).
2. Projections to external collection.

For the first type of projections, it'd be better to use `$set` and store them in the same update as appending events. Then you have the certainty that you won't have race condition and change will be atomic.

For the second part, what you did here should be fine, of course keeping in mind, that process may die in between and those other projections won't be updated, and that we need to apply retries, as mentioned in the first PR.

Reply:
I had the same thought for inline projections, but got a little stuck on aggregating the events that are being appended and getting the correct initial state. Should inline projections _return_ the state to be added to the document? That way the appendToStream func has something to add to the update body or is there a better way to handle that?

Reply:
I think that there are multiple options to handle inline projections:
- return the updated projections object,
- return a Map of updated projections,
- return the updated syntax of MongoDB.

I think that we can start with what's the easiest to support. I would vote to do the following when the projections are registered:
1. Read the projections subdocument (so it contains all inline projections).
2. Run handle methods on top of it. The single inline projection registration would need to contain key of the projection in the projection subdocument. This key could be optional and fall back to "default" if not provided, as my guess is that most of the people would just have one document.
3. Update it in the same change using `$set` operator.

@alex-laycalvert, thoughts?

Reply:
That makes sense, I think the `name` that we register the projection with should suffice as a key, stored in `details.name` on the projection. We could also add a separate key property.

I'll play around with this

Reply:
@alex-laycalvert Sounds good. 👌

It would be also good to double check if our structure is good for indexing strategies. I'm not sure what plays well in Mongo :+1: